### PR TITLE
Ability to control the source of data (test or live) for all page types

### LINKED ISSUE #4822: Control the source of content data between Live/Test for TEST MAP pages
**Is your feature request related to a problem? Please describe.**
We want to be able to control whether we use PROD or TEST content data for MAP pages on the TEST environment. This is going to require additional mechanism being added to Mozart.

**Describe the solution you'd like**
Simorgh will accept a URL param to specify the source environment of content data (e.g. TEST or LIVE) and will pass this to Mozart. This will allow us to load TEST content or LIVE content based on this control parameter.

`renderer_env=int|test|live` e.g. http://www.example.com/pidgin/12345?renderer_env=live

NOTE Live Simorgh should never be able to load TEST content.

**Describe alternatives you've considered**

**Testing notes**
[Tester to complete]

Dev insight: Will Cypress tests be required or are unit tests sufficient? Will there be any potential regression? etc

- [ ] This feature is expected to need manual testing.

**Additional context**

### PR DESCRIPTION
**Pull request by karinathomasbbc**
Resolves #4822 

**Overall change:** 
Users can now provide a parameter in the URL to control where the data will be loaded from.
The parameter name is `renderer_env`
The possible values are: int, test, live
Usage: http://www.example.com/lang/12345?renderer_env=live

**Code changes:**
Updated getUrl to check for presence of renderer_env in the URl.  If it exists, check the value.  If the value is valid, append it to the URL, otherwise ignore.
The valid paths for each article types also needs to support this value being passed in the query string
Adding unit tests

---

- [x] I have assigned myself to this PR and the corresponding issues
- [x] Automated (jest and/or cypress) tests added (for new features) or updated (for existing features)
- [ ] If necessary, I have run the local E2E non-smoke tests relevant to my changes (`CYPRESS_APP_ENV=local CYPRESS_SMOKE=false npm run test:e2e:interactive`) 
- [ ] This PR requires manual testing

### CODE REVIEW

File: src/app/routes/getInitialData/index.js [hunk 1]
```diff
@@ -10,15 +10,39 @@ const logger = nodeLogger(__filename);
...
+  let finalPath = pathname;
+  let renderer = '';
+
+  if (index > 0) {
+    environment = pathname.substring(index + RENDERER_ENV.length);
+
+    if (validRendererEnvironments.includes(environment)) {
+      renderer = `?${RENDERER_ENV}=${environment}`;
+    } else {
+      logger.warn(
```
Comment (karinathomasbbc) [inline src/app/routes/getInitialData/index.js]: Are reviewers happy for this to be at warning?  Or is info/debug better?

File: src/app/routes/getInitialData/index.js [hunk 2]
```diff
@@ -10,15 +10,39 @@ const logger = nodeLogger(__filename);
...
 
 const ampRegex = /(.amp)$/;
 
 const baseUrl = onClient()
   ? getBaseUrl(window.location.origin)
   : process.env.SIMORGH_BASE_URL;
 
-const getUrl = pathname => `${baseUrl}${pathname.replace(ampRegex, '')}.json`;
+export const getUrl = pathname => {
+  const index = pathname.indexOf(RENDERER_ENV) + 1;
```
Comment (ryanmccombe) [inline src/app/routes/getInitialData/index.js]: I see what you mean about the regex!

If you did want a regex approach to this instead of the string stuff, it would be something like:

```
const regex = /renderer_env=(.*)&?/
const [ match, environment ] = pathname.match(regex) || [];
```

File: src/app/routes/getInitialData/index.js [hunk 3]
```diff
@@ -10,15 +10,39 @@ const logger = nodeLogger(__filename);
...
+  let environment;
+  let finalPath = pathname;
+  let renderer = '';
+
+  if (index > 0) {
+    environment = pathname.substring(index + RENDERER_ENV.length);
```
Comment (ryanmccombe) [inline src/app/routes/getInitialData/index.js]: May be worth considering making the querystring case insensitive, eg:
```suggestion
    environment = pathname.substring(index + RENDERER_ENV.length).toLowerCase();
```
and similar handling of `render_env`

**Review by ryanmccombe — DISMISSED** [review]
I'd suggest regex-ing would make this clearer than string-indexing - looks good though!

File: src/app/routes/getInitialData/index.js [hunk 4]
```diff
@@ -6,19 +6,48 @@ import onClient from '#lib/utilities/onClient';
...
 
-const getUrl = pathname => `${baseUrl}${pathname.replace(ampRegex, '')}.json`;
+// eslint-disable-next-line consistent-return
+export const getUrl = pathname => {
+  if (pathname) {
+    const regex = /renderer_env=(.*)&?/i;
+    const [match, environment] = pathname.match(regex) || [];
+    let finalPath = pathname;
+    let params =
+      pathname.lastIndexOf('?') > 0
```
Comment (karinathomasbbc) [inline src/app/routes/getInitialData/index.js]: Should this be indexOf instead?

File: src/app/routes/getInitialData/index.js [hunk 5]
```diff
@@ -6,19 +6,48 @@ import onClient from '#lib/utilities/onClient';
...
 const ampRegex = /(.amp)$/;
 
 const baseUrl = onClient()
   ? getBaseUrl(window.location.origin)
   : process.env.SIMORGH_BASE_URL;
 
-const getUrl = pathname => `${baseUrl}${pathname.replace(ampRegex, '')}.json`;
+// eslint-disable-next-line consistent-return
+export const getUrl = pathname => {
+  if (pathname) {
```
Comment (ryanmccombe) [inline src/app/routes/getInitialData/index.js]: we have no rules around single returns - you can return early if it makes the code cleaner / more performant

eg - you could remove this entire level of nesting and the `eslint-disable` by adding `if (!pathname) return ''` to the first line of the function

File: src/app/routes/getInitialData/index.js [hunk 6]
```diff
@@ -6,19 +6,48 @@ import onClient from '#lib/utilities/onClient';
...
 
 const ampRegex = /(.amp)$/;
 
 const baseUrl = onClient()
   ? getBaseUrl(window.location.origin)
   : process.env.SIMORGH_BASE_URL;
 
-const getUrl = pathname => `${baseUrl}${pathname.replace(ampRegex, '')}.json`;
+// eslint-disable-next-line consistent-return
+export const getUrl = pathname => {
```
Comment (ryanmccombe) [inline src/app/routes/getInitialData/index.js:32]: may want to disable all this in production and fall back to the original implementation - this code runs on every request, and will have no effect when live

eg:
```
const basePath = `${baseUrl}${pathname.replace(ampRegex, '')}.json`;
if (process.env.APP_ENV === 'live') return basePath;
```

or in the interests of removing complexity from this function, leave the original as it is, and switch between them in `fetchData` (line 93-ish)

File: src/app/routes/getInitialData/index.js [hunk 7]
```diff
@@ -6,19 +6,48 @@ import onClient from '#lib/utilities/onClient';
...
+        params = params.replace(match, match.toLowerCase());
+        if (!validRendererEnvironments.includes(environment.toLowerCase())) {
+          logger.warn(
+            `Invalid parameter value [${match}]. Usage: renderer_env=${validRendererEnvironments.join(
+              '|',
+            )}`,
+          );
+        }
+      }
+    }
```
Comment (ryanmccombe) [inline src/app/routes/getInitialData/index.js]: bit more going on here than I thought - wasn't expecting this to effectively allow support for any querystring parameter at all.  Seems okay to do, but I'd recommend addressing the code climate warning re: complexity if we are doing this.  It seems a legitimate warning in this case - there's a lot going on here

eg, line 38-44 being something like `validateRendererEnvironment(environment)` would remove some complexity - other suggestions would be things like `getQuerystring(pathname)` and `getRendererEnvironment(querystring)`

File: src/app/routes/getInitialData/index.js [hunk 8]
```diff
@@ -6,19 +6,48 @@ import onClient from '#lib/utilities/onClient';
...
 const baseUrl = onClient()
   ? getBaseUrl(window.location.origin)
   : process.env.SIMORGH_BASE_URL;
 
-const getUrl = pathname => `${baseUrl}${pathname.replace(ampRegex, '')}.json`;
+// eslint-disable-next-line consistent-return
+export const getUrl = pathname => {
+  if (pathname) {
+    const regex = /renderer_env=(.*)&?/i;
+    const [match, environment] = pathname.match(regex) || [];
```
Comment (ryanmccombe) [inline src/app/routes/getInitialData/index.js]: We have Ramda in simorgh - ramda's `match` function returns an empty array when there are no matches - would be an option here to remove the `||` and associated cyclomatic complexity hit in code climate

https://ramdajs.com/docs/#match

File: src/app/routes/getInitialData/index.test.js [hunk 9]
```diff
@@ -336,4 +339,47 @@ describe('fetchData', () => {
       });
     });
   });
+
+  describe('getUrl', () => {
```
Comment (ryanmccombe) [inline src/app/routes/getInitialData/index.test.js]: really nice tests 👍

File: src/app/routes/getInitialData/index.js [hunk 10]
```diff
@@ -6,19 +6,48 @@ import onClient from '#lib/utilities/onClient';
 import getBaseUrl from './utils/getBaseUrl';
 import getPreprocessorRules from './utils/getPreprocessorRules';
 
-const logger = nodeLogger(__filename);
+export const logger = nodeLogger(__filename);
```
Comment (ryanmccombe) [inline src/app/routes/getInitialData/index.js]: If this is for mocking, we have this: https://github.com/bbc/simorgh/blob/41205ae24deb481e7bfaeb190651188602d631ac/src/testHelpers/loggerMock.js

**Review by ryanmccombe — APPROVED** [review]
Looks good 👍

File: src/app/routes/getInitialData/index.js [hunk 11]
```diff
@@ -6,19 +6,48 @@ import onClient from '#lib/utilities/onClient';
...
 const STATUS_CODE_OK = 200;
 const STATUS_CODE_BAD_GATEWAY = 502;
 const STATUS_CODE_NOT_FOUND = 404;
 const upstreamStatusCodesToPropagate = [STATUS_CODE_OK, STATUS_CODE_NOT_FOUND];
+const validRendererEnvironments = ['int', 'test', 'live'];
```
Comment (chris-hinds) [inline src/app/routes/getInitialData/index.js]: Simorgh doesn't have an `int` environment and Mozart Int points to test simorgh so we can probably remove `int` from the array here.

File: src/app/routes/getInitialData/index.js [hunk 12]
```diff
@@ -6,19 +6,48 @@ import onClient from '#lib/utilities/onClient';
...
 
 const baseUrl = onClient()
   ? getBaseUrl(window.location.origin)
   : process.env.SIMORGH_BASE_URL;
 
-const getUrl = pathname => `${baseUrl}${pathname.replace(ampRegex, '')}.json`;
+// eslint-disable-next-line consistent-return
+export const getUrl = pathname => {
+  if (pathname) {
+    const regex = /renderer_env=(.*)&?/i;
```
Comment (chris-hinds) [inline src/app/routes/getInitialData/index.js]: Could this variable name be a little more descriptive. Maybe something like `rendererEnvRegex`

File: src/app/routes/getInitialData/index.js (hunk 6 above)
Comment (chris-hinds) [inline src/app/routes/getInitialData/index.js:32]: There is a lot going on in this method, and I found it a little hard to follow on the first pass.

Do you think it would be worth trying a library for parsing query strings? This one looks reasonable and is very small (only 452b gzipped) https://www.npmjs.com/package/querystringify

This may or may not be helpful, however it would be worth investigating.

**Review by chris-hinds — COMMENTED** [review]
Nice work 👍 Just left a couple of small comments.

File: src/app/routes/getInitialData/index.js (hunk 6 above)
Comment (karinathomasbbc) [reply src/app/routes/getInitialData/index.js:32]: Thanks for this - querystringify looks super easy to use, so I am going to have a look.

File: src/app/routes/getInitialData/index.js (hunk 8 above)
Comment (karinathomasbbc) [reply src/app/routes/getInitialData/index.js]: Cheers, will have a look

Comment (karinathomasbbc) [reply src/app/routes/getInitialData/index.js]: Using querystringify, regex is no longer needed :)

File: src/app/routes/getInitialData/index.js (hunk 7 above)
Comment (karinathomasbbc) [reply src/app/routes/getInitialData/index.js]: Good shout - there was definitely a lot going on.  I have now made this much simpler.

File: src/app/routes/getInitialData/index.js (hunk 6 above)
Comment (karinathomasbbc) [reply src/app/routes/getInitialData/index.js:32]: I did want to implement this i.e. disabled in production - however, if we do supply a query string, it is included in the pathname then`${baseUrl}${pathname.replace(ampRegex, '')}.json` converts to the following: `www.bbc.com/test/12345?param=test` -> `www.bbc.com/test/12345?param=test.json` which I think is incorrect

**Review by chris-hinds — APPROVED** [review]
Nice work 👍

File: src/app/routes/getInitialData/index.js (hunk 6 above)
Comment (ryanmccombe) [reply src/app/routes/getInitialData/index.js:32]: ah good point - I believe that is the case in the current code too, so you've fixed a bug we don't even know we have yet :D

Comment (jamesdonoh) [reply src/app/routes/getInitialData/index.js:32]: Was just discussing this with Karina, I think we should explicitly prevent overrides from ever working in the live environment in simorgh in case downstream behaviour changes. Ideally we should strip the code from the live bundle entirely but that may be a follow-up task.

Comment (jamesdonoh) [reply src/app/routes/getInitialData/index.js:32]: See @graemelindsayBBC's comment
> Live Simorgh should never be able to load TEST content.

File: src/app/routes/getInitialData/index.js [hunk 13]
```diff
@@ -18,7 +20,47 @@ const baseUrl = onClient()
   ? getBaseUrl(window.location.origin)
   : process.env.SIMORGH_BASE_URL;
 
-const getUrl = pathname => `${baseUrl}${pathname.replace(ampRegex, '')}.json`;
+export const validateRendererEnvironment = environment => {
```
Comment (jroebu14) [inline src/app/routes/getInitialData/index.js]: I'm questioning the importance of logging this user error 🤔 was it a requirement in the issue?

File: src/app/routes/getInitialData/index.js [hunk 14]
```diff
@@ -18,7 +20,47 @@ const baseUrl = onClient()
...
+  const validRendererEnvironments = ['test', 'live'];
+
+  if (
+    !(
+      environment &&
+      validRendererEnvironments.includes(environment.toLowerCase())
+    )
```
Comment (jroebu14) [inline src/app/routes/getInitialData/index.js]: this is hard for me to parse. it helps when things like this are in a well named variable lke:

```js
if (isValidEnv) {
```

I also wouldn't bother correcting any user error casing issues with `environment.toLowerCase()`. I think if we can't type in 'test' or 'live' properly then it's fine not to do anything.

File: src/app/routes/getInitialData/index.js [hunk 15]
```diff
@@ -18,7 +20,47 @@ const baseUrl = onClient()
...
+    logger.warn(
+      `Invalid parameter value [${environment}]. Usage: renderer_env=${validRendererEnvironments.join(
+        '|',
+      )}`,
+    );
+  }
+};
+
+const getPathFromUrl = url => {
+  return url.split(/[?]/)[0];
```
Comment (jroebu14) [inline src/app/routes/getInitialData/index.js]: I think there's a global API for this sort of thing e.g.

```js
const url = new URL('https://developer.mozilla.org/en-US/docs/Web/API/URL')
return url.pathname // '/en-US/docs/Web/API/URL'
```

File: src/app/routes/getInitialData/index.js [hunk 16]
```diff
@@ -18,7 +20,47 @@ const baseUrl = onClient()
...
+};
+
+const checkAndAppendParams = pathname => {
```
Comment (jroebu14) [inline src/app/routes/getInitialData/index.js]: I've not looked at the details of the function yet but the naming suggests a code smell https://github.com/ryanmcdermott/clean-code-javascript#functions-should-do-one-thing

Also, I think `check` in the name doesn't give much context. What does it check for?

File: src/app/routes/getInitialData/index.js [hunk 17]
```diff
@@ -6,6 +6,8 @@ import onClient from '#lib/utilities/onClient';
 import getBaseUrl from './utils/getBaseUrl';
 import getPreprocessorRules from './utils/getPreprocessorRules';
 
+const qs = require('querystringify');
```
Comment (jroebu14) [inline src/app/routes/getInitialData/index.js]:
```suggestion
import qs from 'querystringify';
```

File: src/app/routes/getInitialData/index.js [hunk 18]
```diff
@@ -18,7 +20,47 @@ const baseUrl = onClient()
...
+    );
+  }
+};
+
+const getPathFromUrl = url => {
+  return url.split(/[?]/)[0];
+};
+
+const checkAndAppendParams = pathname => {
+  const url = getPathFromUrl(pathname);
```
Comment (jroebu14) [inline src/app/routes/getInitialData/index.js]: not sure this makes sense. should it not be `const path = getPathFromUrl(url);`?

File: src/app/routes/getInitialData/index.js [hunk 19]
```diff
@@ -18,7 +20,47 @@ const baseUrl = onClient()
...
+  const params = qs.parse(pathname.substring(url.length));
+
+  if (params && params.renderer_env) {
+    params.renderer_env = params.renderer_env.toLowerCase();
```
Comment (jroebu14) [inline src/app/routes/getInitialData/index.js]: We try take an immutable approach when we can - https://wecodetheweb.com/2016/02/12/immutable-javascript-using-es6-and-beyond/ so I wouldn't mutate this object. You also lowercase `renderer_env` here and then lowercase it again in `validateRendererEnvironment`

File: src/app/routes/getInitialData/index.js [hunk 20]
```diff
@@ -18,7 +20,47 @@ const baseUrl = onClient()
...
+    validateRendererEnvironment(params.renderer_env);
+  }
+
+  return `${baseUrl}${url.replace(ampRegex, '')}.json${qs.stringify(
```
Comment (jroebu14) [inline src/app/routes/getInitialData/index.js]: `${baseUrl}${url` doesn't read well. Should be `${baseUrl}${pathname` like it originally was

File: package.json [hunk 21]
```diff
@@ -106,6 +106,7 @@
     "ora": "^4.0.3",
     "path-to-regexp": "^3.1.0",
     "polyfill-crypto.getrandomvalues": "1.0.0",
+    "querystringify": "^2.1.1",
```
Comment (jroebu14) [inline package.json]: I'd always question the value of pulling in new deps as it adds to our bundle size. Is it not too much trouble to live without this package?

File: src/app/routes/getInitialData/index.js (hunk 16 above)
Comment (jroebu14) [reply src/app/routes/getInitialData/index.js]: So this function takes an url and splits out the query params and validates them, if not valid then logs it and then builds the url with query params back up?

**Review by jroebu14 — COMMENTED** [review]
Quite a lot of complexity around just logging an user error (user being a dev). Can we just not do this?

File: src/app/routes/getInitialData/index.js (hunk 13 above)
Comment (jamesdonoh) [reply src/app/routes/getInitialData/index.js]: Yeah, I don't think the logging is really essential here. There are plenty of other things we probably should be logging but aren't!

File: src/app/routes/getInitialData/index.js (hunk 6 above)
Comment (karinathomasbbc) [reply src/app/routes/getInitialData/index.js:32]: Query string params are now removed in live environment

File: src/app/routes/getInitialData/index.js (hunk 13 above)
Comment (karinathomasbbc) [reply src/app/routes/getInitialData/index.js]: Agreed - logging removed!

File: src/app/routes/getInitialData/index.js (hunk 14 above)
Comment (karinathomasbbc) [reply src/app/routes/getInitialData/index.js]: I have removed this functionality, but good shout - something to bear in mind for future - ensuring the code is readable & maintainable in the future.

File: package.json (hunk 21 above)
Comment (karinathomasbbc) [reply package.json]: I am using url-parse (which is already available in node) instead of querystringify

File: src/app/routes/getInitialData/index.js [hunk 22]
```diff
@@ -1,5 +1,6 @@
 import 'isomorphic-fetch';
 import path from 'ramda/src/path';
+import Url from 'url-parse';
```
Comment (ryanmccombe) [inline src/app/routes/getInitialData/index.js:3]: this needs added to package.json (`npm i --save url-parse`)

Comment (chris-hinds) [reply src/app/routes/getInitialData/index.js:3]: good spot

Comment (karinathomasbbc) [reply src/app/routes/getInitialData/index.js:3]: Thanks - will install it now & update.

File: src/app/routes/getInitialData/index.js (hunk 15 above)
Comment (karinathomasbbc) [reply src/app/routes/getInitialData/index.js]: Using url-parse (already installed) works when I don't have the full URL

File: src/app/routes/getInitialData/index.js (hunk 16 above)
Comment (karinathomasbbc) [reply src/app/routes/getInitialData/index.js]: Thanks - this is very useful - reminder to keep my code clean! I have reworked it.

File: src/app/routes/getInitialData/index.js (hunk 17 above)
Comment (karinathomasbbc) [reply src/app/routes/getInitialData/index.js]: `querystringify` no longer required, but thanks for this
